File: android/src/main/java/com/onegini/mobile/OneginiSDK.java
package com.onegini.mobile;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.TimeUnit;

import android.content.Context;

import com.onegini.mobile.view.handlers.CreatePinRequestHandler;
import com.onegini.mobile.view.handlers.PinAuthenticationRequestHandler;
import com.onegini.mobile.view.handlers.RegistrationRequestHandler;
import com.onegini.mobile.sdk.android.client.OneginiClient;
import com.onegini.mobile.sdk.android.client.OneginiClientBuilder;

public class OneginiSDK {

  public static OneginiClient getOneginiClient(final Context context) {
    OneginiClient oneginiClient = OneginiClient.getInstance();
    if (oneginiClient == null) {
      oneginiClient = buildSDK(context);
    }
    return oneginiClient;
  }

  private static OneginiClient buildSDK(final Context context) {
    final Context applicationContext = context.getApplicationContext();
    final RegistrationRequestHandler registrationRequestHandler = new RegistrationRequestHandler(applicationContext);
    final PinAuthenticationRequestHandler pinAuthenticationRequestHandler = new PinAuthenticationRequestHandler(applicationContext);
    final CreatePinRequestHandler createPinRequestHandler = new CreatePinRequestHandler(applicationContext);

    // will throw OneginiConfigNotFoundException if OneginiConfigModel class can't be found
    return new OneginiClientBuilder(applicationContext, createPinRequestHandler, pinAuthenticationRequestHandler)
        // handlers for optional functionalities
        .setBrowserRegistrationRequestHandler(registrationRequestHandler)
        // Set security controller
        .setSecurityController(SecurityController.class)
        // Set http connect / read timeout
        .setHttpConnectTimeout((int) TimeUnit.SECONDS.toMillis(5))
        .setHttpReadTimeout((int) TimeUnit.SECONDS.toMillis(20))
        .build();
  }
}

File: android/src/main/java/com/onegini/mobile/sdk/reactnative/OneginiSDK.kt
package com.onegini.mobile.sdk.reactnative

import android.content.Context
import android.util.Log
import com.onegini.mobile.sdk.android.client.OneginiClient
import com.onegini.mobile.sdk.android.client.OneginiClientBuilder
import com.onegini.mobile.sdk.reactnative.handlers.customregistration.CustomRegistrationActionFactory
import com.onegini.mobile.sdk.reactnative.handlers.fingerprint.FingerprintAuthenticationRequestHandler
import com.onegini.mobile.sdk.reactnative.handlers.mobileauthotp.MobileAuthOtpRequestHandler
import com.onegini.mobile.sdk.reactnative.handlers.pins.CreatePinRequestHandler
import com.onegini.mobile.sdk.reactnative.handlers.pins.PinAuthenticationRequestHandler
import com.onegini.mobile.sdk.reactnative.handlers.registration.RegistrationRequestHandler
import com.onegini.mobile.sdk.reactnative.managers.RegistrationActionManager
import com.onegini.mobile.sdk.reactnative.model.rn.OneginiReactNativeConfig
import com.onegini.mobile.sdk.reactnative.model.rn.ReactNativeIdentityProvider
import com.onegini.mobile.sdk.reactnative.utils.ClassLoader
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class OneginiSDK @Inject constructor(
    private val applicationContext: Context,
    private val registrationRequestHandler: RegistrationRequestHandler,
    private val pinAuthenticationRequestHandler: PinAuthenticationRequestHandler,
    private val createPinRequestHandler: CreatePinRequestHandler,
    private val mobileAuthOtpRequestHandler: MobileAuthOtpRequestHandler,
    private val fingerprintAuthenticationRequestHandler: FingerprintAuthenticationRequestHandler,
    private val customRegistrationFactory: CustomRegistrationActionFactory,
    private val registrationActionManager: RegistrationActionManager,
) {
    var isInitialized = false
        private set

    val oneginiClient: OneginiClient
        get() = OneginiClient.getInstance() ?: buildSDK(applicationContext)

    fun setSDKInitialized() {
        isInitialized = true
    }

    private fun buildSDK(context: Context): OneginiClient {
        val applicationContext = context.applicationContext

        val clientBuilder = OneginiClientBuilder(applicationContext, createPinRequestHandler, pinAuthenticationRequestHandler)

        clientBuilder.setBrowserRegistrationRequestHandler(registrationRequestHandler)
            .setMobileAuthWithOtpRequestHandler(mobileAuthOtpRequestHandler)
            .setFingerprintAuthenticationRequestHandler(fingerprintAuthenticationRequestHandler)
            .setHttpConnectTimeout(Constants.httpConnectTimeoutBrowserRegistrationMiliseconds)
            .setHttpReadTimeout(Constants.httpReadTimeoutBrowserRegistrationMiliseconds)

        val identityProviders = loadIdentityProvidersFromConfig(context)
        addIdentityProviders(identityProviders, clientBuilder)
        return clientBuilder.build()
    }

    private fun loadIdentityProvidersFromConfig(context: Context): List<ReactNativeIdentityProvider> {
        return try {
            val configClass = ClassLoader(context).getClassByName("ReactNativeConfig").newInstance() as OneginiReactNativeConfig
            configClass.getIdentityProviders()
        } catch (e: Exception) {
            Log.e("onegini", "Loading of ReactNativeConfig failed. $e")
            emptyList()
        }
    }

    private fun addIdentityProviders(identityProviders: List<ReactNativeIdentityProvider>, clientBuilder: OneginiClientBuilder) {
        identityProviders.forEach { identityProvider ->
            val provider = customRegistrationFactory.createCustomRegistrationAction(identityProvider)
            registrationActionManager.addCustomRegistrationAction(provider)
            clientBuilder.addCustomIdentityProvider(provider)
        }
    }
}
